Firestore/core/src/remote/grpc_connection.cc

```diff
@@ -71,10 +72,40 @@ std::shared_ptr<grpc::ChannelCredentials> CreateSslCredentials(
   return grpc::SslCredentials(options);
 }
 
-struct HostConfig {
-  util::Path certificate_path;
-  std::string target_name;
-  bool use_insecure_channel = false;
+class HostConfig {
```

Comment:
To reduce verbosity, consider using an approach like:
```cpp
class HostConfig {
 public:
 struct Data {
    util::Path certificate_path;
    std::string target_name;
    bool use_insecure_channel = false;
  };

  // `Modify` is a strawman name.
  template <typename CallbackT>
  void Modify(const CallbackT& callback) {
    std::lock_guard<std::mutex> lock(mutex_);
    callback(data_);
  }
  // You could also have a const function that invokes the callback
  // with a `const Data&`, but that's probably overkill.

 private:
  std::mutex mutex_;
  Data data_;
};

host_config.Modify([&](HostConfig::Data& data) {
  data.certificate_path = certificate_path;
  // ...
});
```

Reply:
I can definitely make that change; however, how are the data members *read* with this new design? Would there be a method that returns a copy of `data_`? Returning a *const reference to* `data_` would still introduce a race condition because the caller would read members of `data_` without holding the lock. Maybe I'm just missing the obvious though!

Reply:
I presumed something like:
```cpp
std::string certificate_path;
host_config.Get([&](const HostConfig::Data& data) {
  certificate_path = data.certificate_path;
});
```
(`Modify` as a function name clearly doesn't work here)

Returning a copy would also work though it seems like the callback-invoking function should cover all use cases.